File: 6.py
import sys
from collections import defaultdict

X = 10
Y = 10

def rotate(guards_dir):
    if guards_dir[0] != 0:
        guards_dir[1] = guards_dir[0]
        guards_dir[0] = 0
    else:
        guards_dir[0] = -guards_dir[1]
        guards_dir[1] = 0

def main():
    grid = [[False for _ in range(Y)] for _ in range(X)]

    with open("Input.txt", "r") as f:
        lines = f.readlines()

    guards_pos = None
    y = 0
    for line in lines:
        x = 0
        for char in line.strip():
            if char == '#':
                grid[x][y] = True
            elif char == '^':
                guards_pos = [x, y]
            else:
                grid[x][y] = False
            x += 1
        y += 1

    guards_dir = [0, -1]
    loops = 0

    while True:
        new_x = guards_pos[0] + guards_dir[0]
        new_y = guards_pos[1] + guards_dir[1]

        if new_x >= X or new_x < 0 or new_y >= Y or new_y < 0:
            break

        if not grid[new_x][new_y]:
            guards_pos[0] = new_x
            guards_pos[1] = new_y
        else:
            tmp_pos = guards_pos[:]
            tmp_dir = guards_dir[:]

            visited = set()
            visited.add((tuple(tmp_pos), tuple(tmp_dir)))

            while True:
                rotate(tmp_dir)
                new_x_tmp = tmp_pos[0] + tmp_dir[0]
                new_y_tmp = tmp_pos[1] + tmp_dir[1]

                if new_x_tmp >= X or new_x_tmp < 0 or new_y_tmp >= Y or new_y_tmp < 0:
                    break

                if not grid[new_x_tmp][new_y_tmp]:
                    tmp_pos[0] = new_x_tmp
                    tmp_pos[1] = new_y_tmp
                else:
                    rotate(tmp_dir)

                state = (tuple(tmp_pos), tuple(tmp_dir))
                if state in visited:
                    loops += 1
                    break

                visited.add(state)

            rotate(guards_dir)
            continue

        rotate(guards_dir)

    print(loops)

if __name__ == "__main__":
    main()

File: 6.cpp
#include <fstream>
#include <iostream>
#include <set>
using namespace std;

int Size = 15;
int X = Size;
int Y = Size;

void Rotate(pair<int, int> &guardsDir) {
    if (guardsDir.first != 0) {
        guardsDir.second = guardsDir.first;
        guardsDir.first = 0;
    } else {
        guardsDir.first = -guardsDir.second;
        guardsDir.second = 0;
    }
}

int main() {
    int map[X][Y];
    ifstream in("Input.txt");
    string lineTmp;
    pair<int, int> guardsPos;
    int y = 0;
    while (getline(in, lineTmp)) {
        int x = 0;
        for (char a : lineTmp) {
            if (a == '#') {
                map[x][y] = 1;
            }
            else if(a == '^') {
                guardsPos.first = x;
                guardsPos.second = y;
                map[x][y] = -1;
            } else {
                map[x][y] = 0;
            }
            x++;
        }
        y++;
    }

    pair<int, int> guardsDir;
    guardsDir.first = 0;
    guardsDir.second = -1;

    int loops = 0;

    while (true) {
        int newX = guardsPos.first + guardsDir.first;
        int newY = guardsPos.second + guardsDir.second;
        if (newX >= X || newX < 0 || newY >= Y || newY < 0) {
            break;
        }
        if (map[newX][newY] != 1) {
            if (map[newX][newY] == 0) {
                auto tmpPos = guardsPos;
                auto tmpDir = guardsDir;

                set<pair<pair<int, int>, pair<int, int>>> visited;
                visited.insert(make_pair(tmpPos, tmpDir));

                map[newX][newY] = 1;

                while (true) {
                    int newXTmp = tmpPos.first + tmpDir.first;
                    int newYTmp = tmpPos.second + tmpDir.second;
                    if (newXTmp >= X || newXTmp < 0 || newYTmp >= Y || newYTmp < 0) {
                        break;
                    }
                    if (map[newXTmp][newYTmp] != 1) {
                        tmpPos.first = newXTmp;
                        tmpPos.second = newYTmp;
                    } else {
                        Rotate(tmpDir);
                    }

                    auto state = std::make_pair(tmpPos, tmpDir);
                    if (visited.contains(state)) {
                        loops++;
                        break;
                    }
                    visited.insert(state);
                }
            }
            map[newX][newY] = -1;
            guardsPos.first = newX;
            guardsPos.second = newY;
            continue;
        }
        Rotate(guardsDir);
    }

    cout << loops;
}
